Decouple CSRF protection from the session machinery

MatthewWilkes opened the pull request:
In #1573 @dstufft makes a number of feature requests about the CSRF implementation of Pyramid, most of which were added in 1.7.

This patch attempts to move the CSRF generation logic out of the session, so new CSRF implementations that do not depend on sessions can be added. It currently does not add any such implementations, the only one shipped is the default which falls back to an implementation of CSRF on the session.

It also introduces new utility functions for `get_csrf_token(request)` and `new_csrf_token(request)` in Python, as well as a `get_csrf_token()` in the template global scope, and repoints documentation to use these rather than explicitly calling the session machinery.

This was written at the DragonSprint by me and @jcerjak, a new contributor to Pyramid.

jcerjak commented on pyramid/config/security.py:
```diff
@@ -218,9 +227,14 @@ def register():
         intr['callback'] = callback
+
+        from pyramid.csrf import csrf_token_template_global
+        from pyramid.events import BeforeRender
```
Could we move these imports to the top?

MatthewWilkes replied:
Done.

MatthewWilkes commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,225 @@
+        csrf = registry.queryUtility(ICSRF)
+        if csrf is not None:
+            event['get_csrf_token'] = partial(csrf.get_csrf_token, request)
+    except AttributeError:
```
This also catches the case when request is None or some weird mock object, like in tests. I've moved the `except` handler only to cover request.registry and the rest in in an `else` handler.

tseaver replied:
The utility promises to implement ICSRF, which guarantees that it has a `get_csrf_token` method.  We don't need to defend against borked implementations.

tseaver commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,225 @@
+)
+from pyramid.interfaces import ICSRF
+from pyramid.settings import aslist
+from zope.interface import implementer
```
Third-party imports should be prior to "local" imports.

tseaver commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,192 @@ which would allow the attacker to control the content of the payload.  Re-using
+
+To use CSRF tokens, you must first enable a :term:`session factory` as
+described in :ref:`using_the_default_session_factory` or
+:ref:`using_alternate_session_factories`.
```
There should be language here which points up the fact that enabling sessions is a requirement of the default `ICSRF` implementation, rather than a general requirement.

MatthewWilkes replied:
I believe @jcerjak is looking at rewording this, is that the case?

jcerjak replied:
I've fixed it in the code docstrings, but missed it here, fixing.

tseaver commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,192 @@ which would allow the attacker to control the content of the payload.  Re-using
+   token = get_csrf_token(request)
+
+The ``get_csrf_token()`` method accepts a single argument, of the request. It
+returns a CSRF *token* string. If ``get_csrf_token()`` or ``new_csrf_token()``
```
> accepts a single argument, of the request.

The "of" should not be there.

tseaver commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,192 @@ which would allow the attacker to control the content of the payload.  Re-using
+
+.. code-block:: javascript
+
+    var csrfToken = ${get_csrf_token()};
```
My JS-fu may be weak, but ISTM that the token value needs to be wrapped in quotes here.

MatthewWilkes replied:
Agreed, the same is true of the current documentation, I didn't pick up on it when I changed `request.session.get_csrf_token` to `get_csrf_token`.

tseaver commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,192 @@ which would allow the attacker to control the content of the payload.  Re-using
+   from pyramid.csrf import get_csrf_token
+   token = new_csrf_token()
+
+Note, it is not possible to force a new CSRF token from a template. If you want
```
Use `.. note::`.

tseaver commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,225 @@
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.get_csrf_token(request)
```
Shouldn't this helper raise if no utility is registered?

MatthewWilkes replied:
I'm not sure, I think you can make arguments either way. If you've written third part code that's intended to be reusable it means you have to special case when users haven't got CSRF enabled.

Either way, we should document the behavior.

mmerickel replied:
> If third-party code registers a view with require_csrf, AFAIK there is no way straightforward way to disable CSRF protection.

This is intentional... if someone defines a view that explicitly requires csrf then it will be inaccessible until someone registers a csrf policy for the app. Turning off csrf protection for those views is a non-goal and can already be done just by re-defining it after `config.commit()` to override the old view.

I'd expect this code to raise if there is no policy. Simply using `getUtility` should be enough.

tseaver commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,225 @@
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.new_csrf_token(request)
```
Shouldn't this helper raise if no utility is registered?

mmerickel replied:
Yes, it should raise. Using `getUtility` should be enough.

tseaver commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,225 @@
+    if supplied_token == "" and header is not None:
+        supplied_token = request.headers.get(header, "")
+
+    impl = request.registry.getUtility(ICSRF)
```
Shouldn't this helper raise if no utility is registered?

MatthewWilkes replied:
As this one is `getUtility` it will raise a `ComponentLookupError` if it's missing, were you thinking of something a bit more descriptive?

tseaver replied:
Ah, my eyes read that as `queryUtility()` like the ones above.

mmerickel commented on docs/whatsnew-1.8.rst:
```diff
@@ -0,0 +1,40 @@
+What's New in Pyramid 1.8
```
Please don't add a whatsnew document. We will do it prior to release. Your changes are good I just think it's easier to create that file from scratch just before a major release.

MatthewWilkes replied:
Got it, thanks!

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,226 @@
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.get_csrf_token(request)
```
Should these be functions or should we add them to the request as methods, similar to `request.has_permission`, `request.session.get_csrf_token` etc? `request.get_csrf_token()`? `request.csrf.get_token()`?

MatthewWilkes replied:
I think that `new_csrf_token` and `get_csrf_token` should be available in the same place, but I'm a little nervous about having `new_csrf_token` available in templates, personally. Maybe I'm worrying over nothing, but it feels to me like something that should only be called in Python code.

bertjwregeer commented:
The CSRF cookie doesn't need to be signed.

bertjwregeer commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,226 @@
+    # should never appear in an URL as doing so is a security issue. We also
+    # explicitly check for request.POST here as we do not support sending form
+    # encoded data over anything but a request.POST.
+    if token is not None:
```
I would recommend swapping the check for the token in the headers/body. `request.POST` is more expensive than checking `request.headers`.

MatthewWilkes replied:
This is unchanged from the previous implementation, but that makes sense. Switching.

bertjwregeer replied:
Thanks @MatthewWilkes!

MatthewWilkes commented:
@mmerickel I've pondered a bit about the change to be specifically about storage, and think that it might be worthwhile to have some of the checking in the CSRF implementation too. Specifically, currently it assumes that constant-time string comparison is sufficient, but an implementation might want to do funky stuff like scope tokens (as mentioned in the original ticket).

If we have a hook there then we can allow this to be supported without having to also supply a `callback`.

MatthewWilkes commented:
@mmerickel @jcerjak and I worked on this more today and incorporated changes based on your feedback. We have added the checking function I referenced in my previous comment, as I think it gives us a bit more flexibility.

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,285 @@
+        return
+    else:
+        csrf = registry.getUtility(ICSRFPolicy)
+        event['get_csrf_token'] = partial(csrf.get_csrf_token, request)
```
This is code from my own apps that I'll at least paste here incase you like it better.

```python
def includeme(config):
    # other session setup usually
    # ...
    config.add_subscriber(renderer_globals, 'pyramid.interfaces.IBeforeRender')

class CSRFToken(object):
    def __init__(self, request):
        self.request = request

    def __json__(self, request):
        return self.request.session.get_csrf_token()

    def __html__(self):
        return self.request.session.get_csrf_token()

def renderer_globals(event):
    request = event['request']
    event['csrf_token'] = CSRFToken(request)  # allows {{ csrf_token }}
```

bertjwregeer replied:
:+1: I like this idea!

jcerjak replied:
I like the simplicity, but maybe it's better to be more explicit here. Don't have a strong opinion about it, but I think it would be better if it's a separate PR.

MatthewWilkes commented:
Will reply to the rest tomorrow, but if you can give us steer on naming that would help. Do you prefer ICSRFStorage?

mmerickel commented:
> Will reply to the rest tomorrow, but if you can give us steer on naming that would help. Do you prefer ICSRFStorage?

Yeah sorry, I know naming is the biggest derailment. Let's go with `ICSRFStoragePolicy` to disambiguate it from the rest of the csrf settings.

MatthewWilkes commented:
@mmerickel With regards to scoped tokens, this isn't included in core (and I personally don't think it should be), but it's nice and easy to implement in userspace with these changes. The implementation would be something like:

```python
from functools import wraps
import hmac

from zope.interface import implementer

from pyramid.csrf import ICSRFStoragePolicy, SessionCSRF
from pyramid.compat import bytes_

def csrf_scope(scope):
    def csrf_scope_decorator(func):
        @wraps(func)
        def with_scope(request, *args, **kwargs):
            request.csrf_scope = scope
            return func(request, *args, **kwargs)
        return with_scope
    return csrf_scope_decorator

@implementer(ICSRFStoragePolicy)
class ScopedSessionCSRF(SessionCSRF):

    def _get_scope(self, request):
        return getattr(request, 'csrf_scope', '')

    def _scope_token(self, scope, base_token):
        token = hmac.new(bytes_(base_token, 'ascii'), scope).hexdigest()
        print "Scoping {} with {} = {}".format(base_token, scope, token)
        return token

    def new_csrf_token(self, request):
        base_token = super(ScopedSessionCSRF, self).new_csrf_token(request)
        return self._scope_token(self._get_scope(request), base_token)

    def get_csrf_token(self, request):
        base_token = super(ScopedSessionCSRF, self).get_csrf_token(request)
        return self._scope_token(self._get_scope(request), base_token)
```

and would be used like:

```python
@view_config(route_name='home', renderer='templates/mytemplate.pt')
@csrf_scope('foobar')
def my_view(request):
    return {'project': 'hacking'}
```

MatthewWilkes commented:
@mmerickel WRT user facing changes, that's a good point. I've added a fallback to the `check_csrf_token` method that ensures that if a user does not configure CSRF the old behaviour of assuming the session will manage it is preserved.

That means that if the user explicitly marks a view as being CSRF checked without calling `config.set_default_csrf_options(require_csrf=True)` they will get SessionCSRF if there is a session present or an error about no session being configured if not.

mmerickel commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,210 @@ which would allow the attacker to control the content of the payload.  Re-using
+   from pyramid.config import Configurator
+
+   config = Configurator()
+   config.set_default_csrf_options(implementation=MyCustomCSRFPolicy())
```
I'd prefer if this were a separate method for `config.set_csrf_storage_policy` as CSRF is entirely unusable without it and it would be good if it could be replaced without changing the other options.

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,291 @@
+
+
+@implementer(ICSRFStoragePolicy)
+class SessionCSRF(object):
```
`SessionCSRFStoragePolicy`

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,291 @@
+        )
+
+@implementer(ICSRFStoragePolicy)
+class CookieCSRF(object):
```
`CookieCSRFStoragePolicy`

mmerickel commented on pyramid/config/security.py:
```diff
@@ -218,9 +231,12 @@ def register():
         intr['safe_methods'] = as_sorted_tuple(safe_methods)
         intr['callback'] = callback
+
+        self.add_subscriber(csrf_token_template_global, [BeforeRender])
```
Currently Pyramid has no precedent for automatically registering a `BeforeRender` subscriber and I think we should keep it that way. If you really want to inject it we could a) do it in the current renderer bindings (pyramid_jinja2, pyramid_mako, pyramid_chameleon) or b) do it in the `pyramid.renderer.RendererHelper` as a `system` value.

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,291 @@
+        self.secure = secure
+        self.httponly = httponly
+        self.domain = domain
+        self.path = path
```
This implementation should be using the `webob.cookies.CookieProfile` from webob in order to set the cookies on potentially multiple domains similar to how `SignedCookieSessionFactory` and `AuthTktAuthenticationPolicy` operate.

MatthewWilkes commented:
@mmerickel I've looked at this again today, as I now have a client who wants CSRF protection without sessions. I've pushed changes to address your comments, let me know when you've had a chance to take a look?

bertjwregeer commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+       considered valid. It must be passed in either the request body or
+       a header.
+
+    .. versionchanged:: 1.8a1
```
We are going to need to change these to Pyramid 1.9

mmerickel commented on pyramid/config/security.py:
```diff
@@ -218,9 +225,25 @@ def register():
+        def register():
+            self.registry.registerUtility(policy, ICSRFStoragePolicy)
+
+        self.action(ICSRFStoragePolicy, register, order=PHASE1_CONFIG)
```
I'm pretty sure the `PHASE1_CONFIG` is unnecessary here. It's used in `set_default_csrf_options` because it needs to be configured prior to `add_view` running the `csrf_view` view deriver. I think the storage policy is only used at request-time though so can be left at the default config phase.

If you think it should be `PHASE1_CONFIG` then it should be added to http://docs.pylonsproject.org/projects/pyramid/en/master/narr/extconfig.html#ordering-actions

MatthewWilkes replied:
I had added it to extconfig, but you're right. Removing it from here and extconfig.

mmerickel commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+        # This means explicit validation has been asked for without configuring
+        # the CSRF implementation. Fall back to SessionCSRFStoragePolicy as that is the
+        # default
+        policy = SessionCSRFStoragePolicy()
```
This is probably unnecessary now. Just use `getUtility` instead an exception if there's no policy. Thoughts?

MatthewWilkes replied:
Yep, thanks. I was missing a call in the testing setup that this was hiding. Fixed.

stevepiercy commented on pyramid/config/views.py:
```diff
@@ -641,16 +641,14 @@ def wrapper(context, request):
-          Note that using this feature requires a :term:`session factory` to
-          have been configured.
-
           .. versionadded:: 1.4a2
```
possibly add a `.. versionchanged 1.x` along with a brief description of the change. I'm not sure into which version these changes will be implemented. @bertjwregeer indicates 1.9 in another comment.

MatthewWilkes replied:
:+1: 

The change is based on the removed text, and adds a glossary term for the CSRF storage policy in the explanation.

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+        return request.session.get_csrf_token()
+
+    def check_csrf_token(self, request, supplied_token):
+        """ Returns True if supplied_token is the same value as get_csrf_token
```
"""Returns ``True`` if ``supplied_token`` is the same value as ``get_csrf_token``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+class CookieCSRFStoragePolicy(object):
+    """ An alternative CSRF implementation that stores its information in
+    unauthenticated cookies, known as the 'Double Submit Cookie' method in the
+    OWASP CSRF guidelines. This gives some additional flexibility with regards
```
Please add link to OWASP resource.

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+        return token
+
+    def check_csrf_token(self, request, supplied_token):
+        """ Returns True if supplied_token is the same value as get_csrf_token
```
"""Returns ``True`` if ``supplied_token`` is the same value as ``get_csrf_token``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+       a header.
+
+    .. versionchanged:: 1.8a1
+       Moved from pyramid.session to pyramid.csrf
```
Moved from :mod:`pyramid.session` to :mod:`pyramid.csrf`.

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+
+def check_csrf_origin(request, trusted_origins=None, raises=True):
+    """
+    Check the Origin of the request to see if it is a cross site request or
```
Not sure why "Origin" is capitalized. Should it be lowercased or have reStructuredText syntax applied?

Also "cross-site".

bertjwregeer replied:
`Origin` is the HTTP Header. All HTTP headers start with an uppercase.

stevepiercy replied:
OK, in that case, it needs some markup:

    ``Origin``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+    Check the Origin of the request to see if it is a cross site request or
+    not.
+
+    If the value supplied by the Origin or Referer header isn't one of the
```
ibid. So perhaps ``Origin`` and ``Referer`` (also ``Referrer``). Although the former misspelling seems to be inherited, I think the correct spelling is valid as well.

bertjwregeer replied:
`Referer` being misspelled is unfortunately in the spec. You will not find that `Referrer` anywhere because it is not valid for an HTTP header.

stevepiercy replied:
``Origin`` or ``Referer``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+
+    If the value supplied by the Origin or Referer header isn't one of the
+    trusted origins and ``raises`` is ``True``, this function will raise a
+    :exc:`pyramid.exceptions.BadCSRFOrigin` exception but if ``raises`` is
```
exception, but

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+    If the value supplied by the Origin or Referer header isn't one of the
+    trusted origins and ``raises`` is ``True``, this function will raise a
+    :exc:`pyramid.exceptions.BadCSRFOrigin` exception but if ``raises`` is
+    ``False`` this function will return ``False`` instead. If the CSRF origin
```
``False``, this

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+    checks are successful this function will return ``True`` unconditionally.
+
+    Additional trusted origins may be added by passing a list of domain (and
+    ports if nonstandard like `['example.com', 'dev.example.com:8080']`) in
```
Let's standardize our "non-standard"s.

Double up the ticks:

    ``['example.com', 'dev.example.com:8080']``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+    (the default) this list of additional domains will be pulled from the
+    ``pyramid.csrf_trusted_origins`` setting.
+
+    Note that this function will do nothing if request.scheme is not https.
```
``request.scheme`` is not ``https``.

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -0,0 +1,283 @@
+    .. versionadded:: 1.7
+
+    .. versionchanged:: 1.8a1
+       Moved from pyramid.session to pyramid.csrf
```
Moved from :mod:`pyramid.session` to :mod:`pyramid.csrf`.

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/interfaces.py:
```diff
@@ -981,19 +981,30 @@ def peek_flash(queue=''):
         """
 
+    def check_csrf_token(request, supplied_token):
+        """ Returns a boolean that represents if supplied_token is a valid CSRF
```
``supplied_token``

MatthewWilkes replied:
:+1:

stevepiercy commented on pyramid/interfaces.py:
```diff
@@ -981,19 +981,30 @@ def peek_flash(queue=''):
+
 class IIntrospector(Interface):
     def get(category_name, discriminator, default=None):
         """ Get the IIntrospectable related to the category_name and the
```
""" Get the ``IIntrospectable`` related to the ``category_name`` and the

mmerickel replied:
@stevepiercy You're commenting on a lot of things that he didn't write in this code. I'm not sure what the right answer is here but I'd rather have him fix the parts he actually wrote and not ask him to do spring cleaning on pre-existing things to get this feature complete.

stevepiercy replied:
I can circle back afterward to do spring cleaning if it doesn't get changed in this round.

MatthewWilkes replied:
I really don't mind spring cleaning lines I've touched, even if they're just moving. I'm applying almost all of @stevepiercy's changed (and, yes, I did see this comment before I started applying them!)

stevepiercy replied:
:hugs:

mmerickel commented:
Just FYI I want to get an alpha of pyramid out soon and this is close enough that I'd rather it didn't miss the cut again. @MatthewWilkes if you don't have time to wrap this up I'll probably do it myself this weekend.

MatthewWilkes commented:
@mmerickel I'll address these issues tomorrow during the work day, thanks for the prod.

MatthewWilkes commented on CHANGES.txt:
```diff
@@ -24,6 +24,14 @@ Features
   appropriately. See https://github.com/Pylons/pyramid/pull/2989
 
+- A new CSRF implementation, :class:`pyramid.csrf.SessionCSRF` has been added,
+  which deleagates all CSRF generation to the current session, following the
```
:+1:

MatthewWilkes commented on docs/narr/security.rst:
```diff
@@ -765,3 +765,210 @@ which would allow the attacker to control the content of the payload.  Re-using
+that it matches one of the trusted origins. By default the only trusted origin
+is the current host, however additional origins may be configured by setting
+``pyramid.csrf_trusted_origins`` to a list of domain names (and ports if they
+are non standard). If a host in the list of domains starts with a ``.`` then
```
Not done as it's old text that's moved, and I don't want to change the corresponding text in the changelog as part of this.

MatthewWilkes commented:
@mmerickel I've pushed changes to the things brought up in this thread, with comments against each either confirming I've made the change (:+1:) or explaining why I haven't.

MatthewWilkes commented on docs/narr/sessions.rst:
```diff
@@ -320,179 +319,3 @@ flash storage.
 .. index::
```
:+1:

bertjwregeer commented:
Thanks @MatthewWilkes!

stevepiercy commented on pyramid/csrf.py:
```diff
@@ -96,8 +96,8 @@ def get_csrf_token(self, request):
     def check_csrf_token(self, request, supplied_token):
-        """ Returns True if supplied_token is the same value as get_csrf_token
-        returns for this request. """
+        """ Returns ``True`` if ``supplied_token is`` the same value as
```
backticks goof:

``supplied_token is``
s/b
``supplied_token`` is

stevepiercy commented:
Thanks @MatthewWilkes. Good stuff. I found only one typo, and then it's good to go from my perspective.

mmerickel commented:
This PR will closed and show as merged automatically when #3019 is merged.

MatthewWilkes commented:
@mmerickel Please don't remove the check_csrf_token method, it is very useful to have it pluggable as that allows for keyring based or fully-stateless CSRF policies.

For example, consider this CSRF policy:

    implementer(ICSRFStoragePolicy)
    class HMACCSRFStoragePolicy(object):
        """ An alternative CSRF implementation that uses short-lived HMAC keyed on
        the user.
        """

        def __init__(self, key):
            self.key = key

        def new_csrf_token(self, request):
            user = request.authenticated_userid
            generated = int(time.time())
            digest = hmac.new(self.key, 'csrf-{}-{}'.format(user, generated), hashlib.sha1).hexdigest()
            return '{}:{}:{}'.format(user, generated, digest)

        def get_csrf_token(self, request):
            return self.new_csrf_token(request)

        def check_csrf_token(self, request, supplied_token):
            user, generated, digest = supplied_token.split(':')
            if user != request.authenticated_userid:
                return False
            if int(generated) < (time.time() - 300):
                return False
            expected = hmac.new(self.key, 'csrf-{}-{}'.format(user, generated), hashlib.sha1).hexdigest()
            return hmac.compare_digest(expected, digest)
            

For non-security experts reading this, I'm not claiming that the above is a good CSRF implementation, however I do think that it demonstrates that there is a value in allowing the CSRF implementation to decide what's valid and what isn't.

If you're concerned about boilerplate (although, personally, I think it's minimal), maybe we could make a mixin available?

bertjwregeer commented:
@MatthewWilkes `get_csrf_token` has access to the `request` object. You will have to manually pull the existing token out of the appropriate header/POST data or wherever, but it is still possible to do what you want to do.

MatthewWilkes commented:
This is true, good point.

mmerickel commented:
@MatthewWilkes thank you very much for the alternative example. I definitely had not considered this stateless use case. It appears that the main downside of supporting that use case without a pluggable `check_csrf_token` is that the policy is now responsible for parsing the header/body itself to find the supplied token. This feels vaguely similar to `unauthenticated_userid` versus `authenticated_userid` in authentication policies. Perhaps I can add another API to get the "unauthenticated" csrf token which can be used if necessary. I'm thinking `find_csrf_token` or `get_unverified_csrf_token`.

MatthewWilkes commented:
Thanks for coming to that decision, it occured to me this morning that Bert's workaround has a fatal flaw. Any submission that includes a CSRF check would cause get_card_token to return the supplied token, making it very easy to hit the short timeout. Imagine a form that takes 3 minutes to fill in, once you've submitted that the resultant page render will use a token with only 2 minutes of validity, so a multipage form would quickly fail.

As to being the most overengineered in the business, I have one word for you, Plone.
